### DISCUSSION TITLE
`outputs.scriptType: 'module'` breaks web-workers

### DISCUSSION THREAD

**Discussion by hevans90** [webpack/webpack#16541] [Q&A]
# Bug report

Changing `outputs.scriptType` from `text/javascript`  to  `module` causes this error when creating web workers with the recommended webpack 5 syntax:

https://webpack.js.org/guides/web-workers/

```typescript
export const worker = new Worker(new URL('./worker.ts', import.meta.url));Ï
```

`Uncaught SyntaxError: Cannot use 'import.meta' outside a module`

I have tried adding ` { type: 'module' }` to the Worker constructor but it does not fix the issue.

**What is the current behavior?**

**If the current behavior is a bug, please provide the steps to reproduce.**

change script type to `module` and witness the error in your console when trying to constuct web workers as follows:

```typescript
export const worker = new Worker(new URL('./worker.ts', import.meta.url));Ï
```

**What is the expected behavior?**

This should not happen.

**Other relevant information:**
webpack version:  
Node.js version: 
Operating System: 
Additional tools:

**Comment by alexander-akait** [comment]
How you load webpack webworkers?

**Reply by hevans90** [reply]
not using a loader, just using webpack 5 as per the docs I linked (the docs say a loader isn't required anymore)

**Reply by alexander-akait** [reply]
I mean what is webpack configuration? `import.meta.url` should be handled by webpack, but looks you didn't get it, that is why I asked  about confoguration, maybe you missing something or override

**Reply by zoran995** [reply]
I have encountered the same issue with the nx monorepo https://github.com/nrwl/nx/issues/28745; perhaps the sample repository shared there may be relevant.

**Reply by alexander-akait** [reply]
@zoran995 I don't think it is webpack problem, looks like a webworker initialization was done invalid, I tried to reproduce locally with without frameworks and everything works fine

Maybe you can create a reproducible test repo with only webpack?

> In some basic tests setting scriptType: "text/javascript" in webpack config makes it work but I am not sure if it might break something else. It actually works even if we override it to undefined; it just fails when it is set to module.

Where did you set it? `scriptType` is `type` for generated `script` (when async chunk loaded)

> p.s. it also works when setting the output module, but in that case,e the HMR doesn't work

Sounds like you generate generate ECMA code, but load it like classic script

**Reply by zoran995** [reply]
Hi, @alexander-akait; thanks for looking into this. Here is a react-webpack repo that reproduces the issue https://github.com/zoran995/ts-react-webpack-worker, it might be easier to figure out what is wrong 

Regarding the HMR, the error is `HookWebpackError: HMR is not implemented for module chunk format yet`, which is being tracked in other tickets, so it’s better not to pollute the discussion here.

**Reply by alexander-akait** [reply]
@zoran995 Why you set `scriptType: "module",`? Your code is not ECMA modules, this option only for `script` tags (which dynamically generated (logic for auto public path), i.e. `import(...)`/`new Worker()`/etc ), I don't think you need this option. Anyway you can fix it using:
```js
plugins: [
    new HtmlWebpackPlugin({
      template: "./src/index.html",
      scriptLoading: "module"
    }),
  ],
```

But if you really want to generate ECMA modules, please use https://webpack.js.org/configuration/output/#outputmodule

**Reply by zoran995** [reply]
The only reason was that Nx did that, and I wanted to check if it is really the source of the issue. I will then try to get them to reopen the issue with the help of details you provided as it can't without output.module set to true and rest details around it

Thanks for looking into this
